src/find_service/server.rs

```diff
@@ -15,73 +12,54 @@ fn convert_system_time_error(time_error: time::SystemTimeError) -> io::Error {
     io::Error::new(io::ErrorKind::Other, time_error)
 }
 
-type PublisherStore = Arc<RwLock<HashMap<String, proto::Registration>>>;
+#[derive(Debug, Clone)]
+pub struct PublisherNotFoundError;
+
+impl fmt::Display for PublisherNotFoundError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid first item to double")
+    }
+}
+
+pub trait PublisherStore: Send + Sync + Clone + 'static {
+    fn insert_publisher(&self, publisher_name: &String, registration: proto::Registration);
+    fn remove_publisher(&self, publisher_name: &String) -> Result<(), PublisherNotFoundError>;
+    fn remove_publishers(&self, publisher_names: &Vec<String>);
```

Comment:
same as with strings, prefer slice references to vec references
```suggestion
    fn remove_publishers(&self, publisher_names: &[String]);
```

You might want to go even further and also allow the `String` type to be flexible:

```rust
    fn remove_publishers<T>(&self, publisher_names: &[T])
    where
        T: Borrow<str>;
```

Reply:
generics might be bad here though, depending on if you need object safety.

Reply:
Not sure it's worth making this generic yet, updated this to use slice refs as suggested though.